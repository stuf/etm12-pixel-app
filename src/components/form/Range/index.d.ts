export interface Tick {
  label: string;
  value: number;
}

export interface Props {
  tickStart: Tick;
  tickEnd: Tick;
  value: number;
  min: number;
  max: number;
  step: number;
  onChange(ev: InputEvent): void;
}
