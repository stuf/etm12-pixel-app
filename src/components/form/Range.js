/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T]"}] */
/**
 * @module Range
 * @namespace components.form
 */
import * as React from 'karet';
import * as U from 'karet.util';
import * as R from 'kefir.ramda';

function Range({
  value,
  min,
  max,
  step = 1,
  onChange,
  className,
  id = 'none',
}) {
  return (
    <div className={U.cns('formField', '--range', className)}>
      <U.Input
        {...{ type: 'range', value, min, max, step, onChange, list: id }}
      />

      <datalist id={id}>
        {U.thru(
          R.range(0, max),
          U.mapElems((it, i) => {
            const isFirst = R.equals(it, min);
            const isLast = R.equals(it, R.subtract(max, 1));
            const label = U.when(R.or(isFirst, isLast), it);

            return (
              <option
                {...{
                  key: U.string`${id}-${i}`,
                  value: it,
                  label,
                }}
              >
                {it}
              </option>
            );
          }),
        )}
      </datalist>
    </div>
  );
}

export default Range;
