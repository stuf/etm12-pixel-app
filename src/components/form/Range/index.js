/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T]"}] */
import * as React from 'karet';
import * as U from 'karet.util';

import * as T from './index.d';
import styles from './index.module.scss';

/**
 * @param {T.Props} props
 */
function Range({ value, min, max, step, onChange }) {
  return (
    <div className={styles.root}>
      <U.Input {...{ type: 'range', value, min, max, step, onChange }} />
    </div>
  );
}

export default Range;
