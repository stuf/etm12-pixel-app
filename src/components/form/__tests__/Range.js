import * as React from 'karet';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Range from 'components/form/Range';

it('renders properly', () => {
  const wrapper = mount(<Range min={0} max={10} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});
