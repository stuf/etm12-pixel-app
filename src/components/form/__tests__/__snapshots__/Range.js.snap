// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<Range
  max={10}
  min={0}
>
  <div
    className="formField --range"
  >
    <ForwardRef(Input)
      list="none"
      max={10}
      min={0}
      step={1}
      type="range"
    >
      <input
        list="none"
        max={10}
        min={0}
        step={1}
        type="range"
      />
    </ForwardRef(Input)>
    <datalist
      id="none"
    >
      <option
        key="none-0"
        label={0}
        value={0}
      >
        0
      </option>
      <option
        key="none-1"
        value={1}
      >
        1
      </option>
      <option
        key="none-2"
        value={2}
      >
        2
      </option>
      <option
        key="none-3"
        value={3}
      >
        3
      </option>
      <option
        key="none-4"
        value={4}
      >
        4
      </option>
      <option
        key="none-5"
        value={5}
      >
        5
      </option>
      <option
        key="none-6"
        value={6}
      >
        6
      </option>
      <option
        key="none-7"
        value={7}
      >
        7
      </option>
      <option
        key="none-8"
        value={8}
      >
        8
      </option>
      <option
        key="none-9"
        label={9}
        value={9}
      >
        9
      </option>
    </datalist>
  </div>
</Range>
`;
